v2/pkg/engine/datasource/graphql_datasource/graphql_subscription_client_test.go

```diff
@@ -2342,6 +2345,218 @@ func TestClientClosesConnectionOnPingTimeout(t *testing.T) {
+			headers:       map[string]string{"Content-Type": "application/json"},
+			expectError:   true,
+			errorContains: "failed to upgrade connection",
+		},
+		{
+			name:          "HTTP 404 Not Found",
+			statusCode:    http.StatusNotFound,
+			headers:       map[string]string{"Content-Type": "text/html"},
+			expectError:   true,
+			errorContains: "failed to upgrade connection",
+		},
+		{
+			name:          "HTTP 500 Internal Server Error",
+			statusCode:    http.StatusInternalServerError,
+			headers:       map[string]string{"Content-Type": "application/json"},
+			expectError:   true,
+			errorContains: "failed to upgrade connection",
+		},
+		{
+			name:          "HTTP 502 Bad Gateway",
+			statusCode:    http.StatusBadGateway,
+			headers:       map[string]string{"Content-Type": "text/html"},
+			expectError:   true,
+			errorContains: "failed to upgrade connection",
+		},
+		{
+			name:          "HTTP 503 Service Unavailable",
+			statusCode:    http.StatusServiceUnavailable,
+			headers:       map[string]string{"Retry-After": "60"},
+			expectError:   true,
+			errorContains: "failed to upgrade connection",
+		},
+		{
+			name:          "HTTP 200 OK (wrong status for WebSocket)",
+			statusCode:    http.StatusOK,
+			headers:       map[string]string{"Content-Type": "application/json"},
+			expectError:   true,
+			errorContains: "failed to upgrade connection",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				for key, value := range tc.headers {
+					w.Header().Set(key, value)
+				}
+				w.WriteHeader(tc.statusCode)
+				fmt.Fprintf(w, `{"error": "WebSocket upgrade failed", "status": %d}`, tc.statusCode)
+			}))
+			defer server.Close()
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			serverCtx, serverCancel := context.WithCancel(context.Background())
+			defer serverCancel()
+
+			client := NewGraphQLSubscriptionClient(http.DefaultClient, http.DefaultClient, serverCtx,
+				WithLogger(logger()),
+			).(*subscriptionClient)
+
+			wsURL := strings.Replace(server.URL, "http://", "ws://", 1)
+
+			updater := &testSubscriptionUpdater{}
+			err := client.Subscribe(resolve.NewContext(ctx), GraphQLSubscriptionOptions{
+				URL: wsURL,
+				Body: GraphQLBody{
+					Query: `subscription {messageAdded(roomName: "room"){text}}`,
+				},
+			}, updater)
+
+			if tc.expectError {
+				require.ErrorContains(t, err, tc.errorContains)
+
+				// Verify the error is of the correct type
+				var upgradeErr *UpgradeRequestError
+				require.ErrorAs(t, err, &upgradeErr)
+				require.Equal(t, tc.statusCode, upgradeErr.StatusCode)
+				require.Equal(t, server.URL, upgradeErr.URL)
+			} else {
+				assert.NoError(t, err, "Expected no error for status code %d", tc.statusCode)
+			}
+
+			// Give some time for any async cleanup
+			time.Sleep(50 * time.Millisecond)
```

Comment:
Any way we can wait on this with a channel and timeout instead of sleeping?

Reply:
ah. removing that since we are not doing leaks detection here.